crates/ra_assists/src/handlers/change_return_type_to_result.rs

```diff
@@ -0,0 +1,798 @@
+                    arm_list.arms().filter_map(|match_arm| match_arm.expr()).for_each(|expr| {
+                        if let Some(block_expr) = BlockExpr::cast(expr.syntax().clone()) {
+                            if let Some(block) = block_expr.block() {
+                                self.collect_return_exprs(&block);
+                            }
+                        } else if let Some(ret_expr) = ast::ReturnExpr::cast(expr.syntax().clone())
+                        {
+                            if let Some(ret_expr_arg) = &ret_expr.expr() {
+                                self.exprs_to_wrap.push(ret_expr_arg.syntax().clone());
+                            }
+                        }
+                    });
+                }
+            }
+            _ => {}
+        }
+    }
+
+    fn collect_tail_exprs(&mut self, block: &Block) {
+        if let Some(expr) = block.expr() {
+            self.fetch_tail_exprs(&expr);
+        }
+    }
+
+    fn fetch_tail_exprs(&mut self, expr: &Expr) {
+        match get_tail_expr_from_block(expr) {
+            Some(exprs) => {
+                for expr in &exprs {
+                    if let Some(last_child) = &expr.last_child() {
+                        if let Some(last_expr) = &Expr::cast(last_child.clone()) {
+                            self.fetch_tail_exprs(last_expr);
+                        } else if let Some(block) = ast::Block::cast(last_child.clone()) {
+                            if let Some(expr) = &block.expr() {
+                                self.fetch_tail_exprs(expr);
+                            }
+                        } else {
+                            self.exprs_to_wrap.push(last_child.clone());
+                        }
+                    } else {
+                        self.exprs_to_wrap.push(expr.clone());
+                    }
+                }
+            }
+            _ => {
+                self.exprs_to_wrap.push(expr.syntax().clone());
+            }
+        }
+    }
+}
+
+/// Get a tail expression inside a block
+fn get_tail_expr_from_block(expr: &Expr) -> Option<Vec<SyntaxNode>> {
+    match expr {
+        Expr::IfExpr(if_expr) => {
+            let mut nodes = vec![];
+            for block in if_expr.blocks() {
+                if let Some(in_block) = block.block() {
+                    if let Some(last_expr) = in_block.syntax().last_child() {
+                        nodes.push(last_expr);
+                    } else {
+                        nodes.push(in_block.syntax().clone());
+                    }
+                } else {
+                    nodes.push(block.syntax().clone());
+                }
+            }
+            Some(nodes)
+        }
+        Expr::LoopExpr(loop_expr) => loop_expr.syntax().last_child().map(|lc| vec![lc]),
+        Expr::ForExpr(for_expr) => for_expr.syntax().last_child().map(|lc| vec![lc]),
+        Expr::WhileExpr(while_expr) => while_expr.syntax().last_child().map(|lc| vec![lc]),
```

Comment:
for these loops, the last expression isn't really special -- we don't want to turn
```rust
loop {
    foo();
    bar()
}
```
into
```rust
loop {
    foo();
    Ok(bar())
}
```